h/static/scripts/base/controller.js

```diff
@@ -0,0 +1,69 @@
+'use strict';
+
+var dom = require('../util/dom');
+
+/**
+ * Base class for controllers that upgrade elements with additional
+ * functionality.
+ *
+ * - Child elements with `js-$name` class prefixes are exposed on the controller
+ *   as `this.refs.$name` where $name is camelCased.
+ * - The element passed to the controller is exposed via the `element` property
+ * - The controllers attached to an element are accessible via the
+ *   `element.controllers` array
+ *
+ * The controller maintains internal state in `this.state`, which can only be
+ * updated by calling (`this.setState(changes)`). Whenever the internal state of
+ * the controller changes, `this.update()` is called to sync the DOM with this
+ * state.
+ *
+ * @param {Element} element - The DOM Element to upgrade
+ */
+function Controller(element) {
+  if (!element.controllers) {
+    element.controllers = [this];
+  } else {
+    element.controllers.push(this);
+  }
+
+  this.state = {};
+  this.element = element;
+  this.refs = dom.jsElements(element);
```

Comment:
I wondered about this, although that could also be confused with referring to the element's direct children. The terminology (`setState`, `refs`) is borrowed from React in a bid to hopefully be a) familiar (at least to some front-end devs) and b) concise. We use the same `setState` pattern in a few places in the client and browser extension.